### ISSUE TITLE
Seg fault during covariance matrix construction for data assimilation only for particular numbers of MPI tasks

### ISSUE THREAD

**Issue by stvdwtt** [adamantine-sim/adamantine#314]
Summary: An IMTS gear simulation crashes after data assimilation only for the case where 27 MPI processes are used.

Test case:
https://github.com/adamantine-sim/demonstration-cases/tree/main/IMTS_Parts/GEAR-OP04

To run this, copy the input files out of `simulation_template` into its parent `GEAR-OP04` directory and then run adamantine.

We get a seg fault at this line:
https://github.com/adamantine-sim/adamantine/blob/587a4ead79730ea7b30829346766f7e8bc781598/source/DataAssimilator.cc#L581

with an index that is greater than the number of degrees of freedom.

It's unclear whether the issue is in `indices_ranks` or if `j` has an invalid value.

@Rombur, could you take a look at this?

**Comment by Rombur** [comment]
Ughh it passed on my machine. My guess is that due to numerical error, ArborX is giving us results we cannot deal with. Have you tried running in debug mode?

**Comment by stvdwtt** [comment]
Yeah, debug mode is how I localized it. It fails with 27 ranks on the SCOPS-foundry-2 computer at the MDF

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Bruno Turcksin ***@***.***>
Sent: Monday, September 9, 2024 2:32:58 PM
To: adamantine-sim/adamantine ***@***.***>
Cc: DeWitt, Stephen ***@***.***>; Author ***@***.***>
Subject: [EXTERNAL] Re: [adamantine-sim/adamantine] Seg fault during covariance matrix construction for data assimilation only for particular numbers of MPI tasks (Issue #314)


Ughh it passed on my machine. My guess is that due to numerical error, ArborX is giving us results we cannot deal with. Have you tried running in debug mode?
